stream/stream.go

```diff
@@ -8,6 +8,186 @@ import (
 type Event struct {
 	Ctx        context.Context
 	Key, Value []byte
+	Ack        func()
+}
+
+type Publisher interface {
+	Subscribe(context.Context, Subscriber, ...SubscriptionOption)
+}
+
+type Subscriber interface {
+	OnNext(*Event) error
+	OnError(error) // called at most once
+	OnComplete()   // called at most once if the subscription is not canceled before the stream completes
+}
+
+type SubscriptionOption func(options SubscriptionConfig)
+
+type subscriber struct {
+	onNext     func(*Event) error
+	onError    func(error)
+	onComplete func()
+}
+
+func (s subscriber) OnNext(event *Event) error {
+	return s.onNext(event)
+}
+
+func (s subscriber) OnError(err error) {
+	s.onError(err)
+}
+
+func (s subscriber) OnComplete() {
+	s.onComplete()
+}
+
+func CreateSubscriber(onNext func(*Event) error, onError func(error), onComplete func()) Subscriber {
+	return subscriber{
+		onNext:     onNext,
+		onError:    onError,
+		onComplete: onComplete,
+	}
+}
+
+// unfortunately we need this to be an interface so that we can extend it for transport specific configuration
+type SubscriptionConfig interface {
```

Comment:
I'm OK to have it as an interface. But could we try with the minimal amount of methods for now?

And should we have the Set in the interface? Shouldn't it only be inside the concrete type?